src/backend/distributed/transaction/multi_transaction.c

```diff
@@ -0,0 +1,1907 @@
+
+	FinishConnectionEstablishment(connection);
+	return connection;
+}
+
+
+static void
+AdjustRemoteTransactionStates(List *connectionList)
+{
+	ListCell *connectionCell = NULL;
+
+	if (!InCoordinatedTransaction())
+		return;
+	foreach(connectionCell, connectionList)
+	{
+		MultiConnection *connection = (MultiConnection *) lfirst(connectionCell);
+
+		Assert(PQtransactionStatus(connection->conn) != PQTRANS_ACTIVE);
+
+		if ((connection->transactionState & MULTI_TRANSACTION_STATE_FAILED) ||
+			(connection->transactionState & MULTI_TRANSACTION_STATE_STARTING) ||
+			(connection->transactionState & MULTI_TRANSACTION_STATE_STARTED))
+		{
+			continue;
+		}
+
+		if (PQtransactionStatus(connection->conn) != PQTRANS_INTRANS)
+		{
+			/*
+			 * Check whether we're right now allowed to start new client
+			 * transaction.  FIXME: This likely can be removed soon.
+			 */
+			if (XactModificationLevel > XACT_MODIFICATION_NONE)
+			{
+				ereport(ERROR, (errcode(ERRCODE_ACTIVE_SQL_TRANSACTION),
+								errmsg("cannot open new connections after the first modification "
+									   "command within a transaction")));
+			}
+
+			/* FIXME: specify isolation level etc. */
+			if (!SendRemoteCommand(connection, "BEGIN;"))
+			{
+				ReportConnectionError(connection, WARNING);
+				connection->transactionState |= MULTI_TRANSACTION_STATE_FAILED;
+			}
+			else
+			{
+				connection->transactionState |= MULTI_TRANSACTION_STATE_STARTING;
+			}
+		}
+	}
+
+
+	foreach(connectionCell, connectionList)
+	{
+		MultiConnection *connection = (MultiConnection *) lfirst(connectionCell);
+		PGresult *result = NULL;
+
+		if (connection->transactionState & MULTI_TRANSACTION_STATE_FAILED)
+		{
+			continue;
+		}
+
+		if (!(connection->transactionState & MULTI_TRANSACTION_STATE_STARTING))
+		{
+			continue;
+		}
+
+		result = PQgetResult(connection->conn);
+		if (!IsResponseOK(result))
+		{
+			connection->transactionState |= MULTI_TRANSACTION_STATE_FAILED;
+			connection->transactionState &= ~MULTI_TRANSACTION_STATE_STARTING;
+			ReportResultError(connection, result, WARNING);
+		}
+		else
+		{
+			connection->transactionState &= ~MULTI_TRANSACTION_STATE_STARTING;
+			connection->transactionState |= MULTI_TRANSACTION_STATE_STARTED;
+		}
+
+		PQclear(result);
+
+		ForgetResults(connection);
+
+		Assert(PQtransactionStatus(connection->conn) == PQTRANS_INTRANS);
+	}
+}
+
+
+void
+AdjustRemoteTransactionState(MultiConnection *connection)
+{
+	if (InCoordinatedTransaction())
+	{
+		List *connectionList = list_make1(connection);
+
+		AdjustRemoteTransactionStates(connectionList);
+		list_free(connectionList);
+	}
+}
+
+/* -------------------------------------------------------------------------
+ * Execution related functions.
+ * -------------------------------------------------------------------------
+ */
+
+/*
+ * Execute commands on placements in parallel.
+ *
+ * TODO: Use less than one one connection per placement.
```

Comment:
I guess what makes it a bit tricky is that when the DDL command is the first modification then you can wait for any connection to become available, but otherwise you need to wait for a specific connection to become available. In the implementation, we probably need some per-connection queue for modifications with a designated connection and possibly some "remaining" queue for others. If a connection becomes available a new task is started from its own queue if any, and otherwise the remaining queue. An alternative would be to assign all modifications to connections upfront and only have a per-connection queue.